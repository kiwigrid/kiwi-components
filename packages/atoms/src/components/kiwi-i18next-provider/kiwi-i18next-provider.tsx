import {
  Component,
  Host,
  h,
  ComponentWillLoad,
  Prop,
  Watch,
  Event,
  EventEmitter,
} from '@stencil/core';
import i18next, { TFunction } from 'i18next';
import i18nextXhrBackend, { BackendOptions } from 'i18next-xhr-backend';
import store from './kiwi-i18next-provider.store';

@Component({
  tag: 'kiwi-i18next-provider',
  shadow: false,
})
export class KiwiI18nextProvider implements ComponentWillLoad {
  /**
   * The language to use for i18n. en as default.
   */
  @Prop()
  public lng = 'en';

  /**
   * Base path used to configure i18next backend.
   *
   * @see {BackendOptions#loadPath}
   */
  @Prop()
  public loadBasePath!: string;

  /** Namespaces to be loaded by i18next */
  @Prop()
  public ns: string[] = ['common'];

  /**
   * This event is dispatched when i18nexts t function changes
   * The t function is passed as data
   */
  @Event()
  private tFunctionChanged!: EventEmitter<TFunction>;

  @Watch('lng')
  async onLanguageChange(newLng: string): Promise<void> {
    const t = await i18next.changeLanguage(newLng);
    store.set('t', t);
    this.tFunctionChanged.emit(store.get('t'));
  }

  public async componentWillLoad(): Promise<void> {
    const backend: BackendOptions = {
      loadPath: () => `${this.loadBasePath}/{{lng}}/{{ns}}.json`,
      crossDomain: true,
    };

    const t = await i18next.use(i18nextXhrBackend).init({
      lng: this.lng,
      fallbackLng: 'en',
      // debug: true,
      ns: this.ns,
      defaultNS: 'common',
      backend,
    });

    store.set('t', t);
    this.tFunctionChanged.emit(store.get('t'));
  }

  render(): JSX.Element {
    return (
      <Host>
        <slot></slot>
      </Host>
    );
  }
}
