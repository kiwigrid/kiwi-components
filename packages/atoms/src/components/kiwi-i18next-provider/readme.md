# kiwi-i18next-provider



<!-- Auto Generated Below -->


## Properties

| Property                    | Attribute        | Description                                  | Type       | Default      |
| --------------------------- | ---------------- | -------------------------------------------- | ---------- | ------------ |
| `i18next` _(required)_      | --               | Uninitialized i18next instance to use.       | `i18n`     | `undefined`  |
| `lng`                       | `lng`            | The language to use for i18n. en as default. | `string`   | `'en'`       |
| `loadBasePath` _(required)_ | `load-base-path` | Base path used to configure i18next backend. | `string`   | `undefined`  |
| `ns`                        | --               | Namespaces to be loaded by i18next           | `string[]` | `['common']` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
