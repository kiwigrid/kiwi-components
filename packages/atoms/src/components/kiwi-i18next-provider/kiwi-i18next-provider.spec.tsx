import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import i18next from 'i18next';
import { KiwiI18nextProvider } from './kiwi-i18next-provider';
import store from './kiwi-i18next-provider.store';

describe('kiwi-i18next-provider', () => {
  beforeEach(() => {
    store.dispose();
  });

  it('renders', async () => {
    const page = await newSpecPage({
      components: [KiwiI18nextProvider],
      template: () => (
        <kiwi-i18next-provider
          i18next={i18next}
          lng="de"
          loadBasePath=""
        ></kiwi-i18next-provider>
      ),
    });

    expect(page.root).toMatchSnapshot();
  });

  it('reacts to changes', async () => {
    const page = await newSpecPage({
      components: [KiwiI18nextProvider],
      template: () => (
        <kiwi-i18next-provider
          i18next={i18next}
          lng="de"
          loadBasePath=""
        ></kiwi-i18next-provider>
      ),
    });

    expect(store.get('t')('common:hello')).toEqual('Hallo');

    (page.root as HTMLKiwiI18nextProviderElement).lng = 'en';

    expect(store.get('t')('common:hello')).toEqual('Hello');
  });
});
