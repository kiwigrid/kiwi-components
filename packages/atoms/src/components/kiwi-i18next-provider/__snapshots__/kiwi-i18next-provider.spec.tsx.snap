// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-i18next-provider renders 1`] = `
<kiwi-i18next-provider>
  <!---->
</kiwi-i18next-provider>
`;
